scipy/optimize/_numdiff.py

```diff
@@ -516,7 +516,19 @@ def _dense_difference(fun, x0, f0, h, use_one_sided, method):
         else:
             raise RuntimeError("Never be here.")
 
-        J_transposed[i] = df / dx
+        if np.count_nonzero(dx == 0.0):
```

Comment:
Are there any other circumstances for which `dx` could be zero - ones for which we'd prefer an error to be raised rather than possibly getting an erroneous result?

Reply:
There are circumstances where `dx` could be zero. For example, the user could call [`approx_fprime`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.approx_fprime.html) with entries in `epsilon` of 0. So it's not just `minimize` methods that call `approx_derivative`.

It's not beyond the realms of possibility that an `fmin_*` or `minimize` method could specify a zero `eps` or a zero `finite_diff_rel_step` (e.g. https://docs.scipy.org/doc/scipy/reference/optimize.minimize-lbfgsb.html#optimize-minimize-lbfgsb). It would be unusual, but not beyond the realms of possibility, that someone would want to do that on purpose.

I did wondered about factoring out the redundant values in `ScalarFunction`. However, that's more involved surgery. The current PR fixes the bugs being reported, and isn't too drastic a change. Factoring out the redundant parameters in ScalarFunction would be a good next step, for the performance increase you mention. However, I suspect cases where `lb==ub` aren't in the majority of problems.

Reply:
So for these other circumstances where `dx` is zero - is it appropriate to return `df/dx = 0` for these cases?

Reply:
Hmm. It's impossible to calculate a finite difference gradient if dx=0. So the gradient should either be 0 or NaN. The latter creates issues for all the `minimize` methods with bounds. But the former might be what is wanted for `approx_fprime`.
Conversion of NaN --> 0 could happen at a higher level in `ScalarFunction`, but some `np.seterr` would need to happen to prevent warnings. That filtering would be better off in lower level code.

Reply:
OK. I just think that if we go this route, we only want to set `df/dx=0`  for the cases where `lb=ub`. We don't want to set `df/dx=0` where `dx=0` for all other reasons. I'm suggesting that we replace the `np.count(dx==0)` with something like `(lb == ub).any()` and mask based on `lb==ub` rather than `dx=0`. Does that make sense?

Reply:
I think I'm leaning towards the other route at the moment.